File: src-mobile/android/core/domain/src/main/kotlin/com/rainif/doneat/core/domain/salary/NumberInput.kt
package com.rainif.doneat.core.domain.salary

/** Number-pad input. Normalizing digits/separators must never remove a sign or truncate an amount. */
object NumberInput {
    /** Null means the entire edit is invalid. Empty and a lone separator are editable drafts. */
    fun normalize(text: String, decimal: Boolean, maxDigits: Int): String? {
        val folded = StringBuilder()
        var digits = 0
        var seenSeparator = false
        val points = text.codePoints().iterator()
        while (points.hasNext()) {
            val c = points.nextInt()
            val digit = Character.digit(c, 10)
            when {
                digit in 0..9 && Character.isDigit(c) -> {
                    if (++digits > maxDigits) return null
                    folded.append(('0' + digit))
                }
                decimal && c in listOf('.'.code, ','.code, '٫'.code) && !seenSeparator -> {
                    seenSeparator = true
                    folded.append('.')
                }
                else -> return null
            }
        }
        return folded.toString()
    }

    /** Invalid text stays visible so it can be corrected; it must not be committed. */
    fun draft(text: String, decimal: Boolean, maxDigits: Int): String = normalize(text, decimal, maxDigits) ?: text

    /** Empty clears the amount; zero remains a real value. Invalid drafts never reach settings. */
    fun committedText(text: String, decimal: Boolean, maxDigits: Int): String? =
        text.takeIf { it == normalize(it, decimal, maxDigits) && (it.isEmpty() || parse(it) != null) }

    /** A canonical decimal value, excluding signs, exponents, non-finite values and partial drafts. */
    fun parse(text: String): Double? = text.takeIf {
        it.isNotEmpty() && it != "." && it == normalize(it, decimal = true, maxDigits = Int.MAX_VALUE)
    }?.toDoubleOrNull()?.takeIf { it.isFinite() && it >= 0 }
}

File: src-mobile/ios/App/App/Native/Models/NumberInput.swift
import Foundation

/// Number-pad input. Normalizing digits/separators must never remove a sign or truncate an amount.
nonisolated enum NumberInput {
    /// Nil rejects the entire edit. Empty and a lone separator are editable drafts.
    static func normalize(_ text: String, decimal: Bool, maxDigits: Int) -> String? {
        var folded = ""
        var digits = 0
        var seenSeparator = false
        for scalar in text.unicodeScalars {
            if scalar.properties.numericType == .decimal,
               let value = Character(String(scalar)).wholeNumberValue, (0...9).contains(value) {
                digits += 1
                guard digits <= maxDigits else { return nil }
                folded += String(value)
            } else if decimal, [".", ",", "\u{066B}"].contains(String(scalar)), !seenSeparator {
                seenSeparator = true
                folded += "."
            } else {
                return nil
            }
        }
        return folded
    }

    /// Invalid text stays visible so it can be corrected; it must not be committed.
    static func draft(_ text: String, decimal: Bool, maxDigits: Int) -> String {
        normalize(text, decimal: decimal, maxDigits: maxDigits) ?? text
    }

    /// Empty clears the amount; zero remains a real value. Invalid drafts never reach settings.
    static func committedText(_ text: String, decimal: Bool, maxDigits: Int) -> String? {
        guard text == normalize(text, decimal: decimal, maxDigits: maxDigits),
              text.isEmpty || parse(text) != nil else { return nil }
        return text
    }

    /// Canonical decimal values exclude signs, exponents, non-finite values and partial drafts.
    static func parse(_ text: String) -> Double? {
        guard !text.isEmpty, text != ".",
              text == normalize(text, decimal: true, maxDigits: .max),
              let value = Double(text), value.isFinite, value >= 0 else { return nil }
        return value
    }
}
